File: sphere/hw_sphere1/task_1_3.py
from math import factorial

n = int(input())
k = int(input())
k -= 1
was = [0]*21
per = [0]*21
for i in range(1, n + 1):
    fac = factorial(n - i)
    a = k // fac
    k %= fac
    cur = 0
    for j in range(1, n + 1):
        if was[j] == 0:
            cur += 1
            if cur == (a + 1):
                per[i] = j
                was[j] = 1
for i in range(1, n + 1):
    print(per[i], end=' ')

File: sphere/hw_sphere1/task1_3.c
#include <stdio.h>
#define length 4


int factorial(int count){
    int v = 1;
    for (int i = 1;i <= count; i++)
        v = v * i;
    return v;
}

int main(void){
    long long int n, k;
    scanf("%lld%lld", &n, &k);
    k -= 1;
    long long int a, cur, fac;
    long long int per[21];
    long long int was[21] = {0};
    if (k > factorial(n)){
        for (long long int i = 1; i <= n; i++){
            printf("%lld ", i);
        }
    } else {
        for (long long int i = 1; i <= n; i++) {
            fac = factorial((n - i));
            a = k / fac;
            k %= fac;
            cur = 0;
            for (long long int j = 1; j <= n; j++) {
                if (was[j] == 0) {
                    cur += 1;
                    if (cur == (a + 1)) {
                        per[i] = j;
                        was[j] = 1;
                    }
                }
            }
        }
        for (long long int i = 1; i <= n; i++) {
            printf("%lldd ", per[i]);
        }
    }
    return 0;
}